### PR TITLE
Fix polygon api

### PR DESCRIPTION
**Pull request by gagiuntoli**
## To-do list

- [ ] Added tests
- [ ] Tested on dev/sbx
- [ ] Changelog was updated with current changes
- [ ] Documentation or guides were updated (slab, readme)
- [ ] Version was updated on `package.json` (releases/hotfixes)

## Checklist before merge

- [ ] **there are no errors in logs in any workers**
- [ ] new blocks are being discovered and marked as finalized
- [ ] foreign blocks are being dispatched
- [ ] squash all commits before merging

### CODE REVIEW

File: src/services/fetchers/MetalPriceApiFetcher.ts [hunk 1]
```diff
@@ -51,7 +51,7 @@ export default class MetalPriceApiFetcher implements FeedFetcherInterface {
         throw new Error(response.data);
       }
 
-      const rate = response.data.rates[currency];
+      const rate = response.data?.rates[symbol];
```
Comment (DZariusz) [inline src/services/fetchers/MetalPriceApiFetcher.ts:54]: why this change?

File: src/services/fetchers/PolygonIOCryptoPriceFetcher.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
 import {inject, injectable} from 'inversify';
 import {Logger} from 'winston';
 
-import PolygonIOCryptoPriceService from '../PolygonIOCryptoPriceService.js';
```
Comment (DZariusz) [inline src/services/fetchers/PolygonIOCryptoPriceFetcher.ts:4]: I see a lot of unrelated changes like this, is your IDE changing some files automatically?

File: src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts [hunk 3]
```diff
@@ -21,15 +25,36 @@ class PolygonIOCurrencySnapshotGramsFetcher extends BasePolygonIOSingleFetcher {
     this.valuePath = '$.ticker.lastQuote.a';
   }
 
-  async apply(params: {ticker: string}): Promise<number> {
+  async apply(params: {ticker: string}, options: FeedFetcherOptions): Promise<number> {
```
Comment (DZariusz) [inline src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts:28]: looks like this was missed because we not have intefrace here, please apply interface for this fetcher

File: src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts [hunk 4]
```diff
@@ -21,15 +25,36 @@ class PolygonIOCurrencySnapshotGramsFetcher extends BasePolygonIOSingleFetcher {
...
+    if (!timestamp || timestamp <= 0) throw new Error(`${this.logPrefix} invalid timestamp value: ${timestamp}`);
+
     this.logger.debug(`${this.logPrefix} call for ${ticker}`);
 
     const data = await this.fetch(sourceUrl);
     const oneOzInGrams = 31.1034; // grams
-    return (data as number) / oneOzInGrams;
+    const price = (data as number) / oneOzInGrams;
+
+    if (price !== null) {
```
Comment (DZariusz) [inline src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts:41]:
```suggestion
    if (price !== NaN) {
```

I think it is not possible for price to be null

File: src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts [hunk 5]
```diff
@@ -21,15 +25,36 @@ class PolygonIOCurrencySnapshotGramsFetcher extends BasePolygonIOSingleFetcher {
...
     const sourceUrl = `https://api.polygon.io/v2/snapshot/locale/global/markets/forex/tickers/${ticker}?apiKey=${this.apiKey}`;
 
+    if (!timestamp || timestamp <= 0) throw new Error(`${this.logPrefix} invalid timestamp value: ${timestamp}`);
+
     this.logger.debug(`${this.logPrefix} call for ${ticker}`);
 
     const data = await this.fetch(sourceUrl);
     const oneOzInGrams = 31.1034; // grams
-    return (data as number) / oneOzInGrams;
+    const price = (data as number) / oneOzInGrams;
```
Comment (DZariusz) [inline src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts:39]:
```suggestion
    const price = parseFloat(data) / oneOzInGrams;
```

File: src/services/fetchers/MetalPriceApiFetcher.ts (hunk 1 above)
Comment (gagiuntoli) [reply src/services/fetchers/MetalPriceApiFetcher.ts:54]: It was wrong, probably during the refactoring I changed it wrongly. Now, it works fine.

File: src/services/fetchers/PolygonIOCryptoPriceFetcher.ts (hunk 2 above)
Comment (gagiuntoli) [reply src/services/fetchers/PolygonIOCryptoPriceFetcher.ts:4]: No, I changed manually because I looked at those files. Every little improvement I can do, I try it to do.

File: src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts (hunk 4 above)
Comment (gagiuntoli) [reply src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts:41]: correct

**Comment by DZariusz** [comment]
please add info to changelog

File: src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts (hunk 5 above)
Comment (gagiuntoli) [reply src/services/fetchers/PolygonIOCurrencySnapshotGramsFetcher.ts:39]: this is not possible right away without modifying extra code, since the function `fetch` returns this:

```typescript
async fetch(sourceUrl: string, raw = false): Promise<SinglePriceResponse | number>
```
and it is not possible to convert that to a string directly.

File: src/services/fetchers/PolygonIOCryptoPriceFetcher.ts (hunk 2 above)
Comment (DZariusz) [reply src/services/fetchers/PolygonIOCryptoPriceFetcher.ts:4]: what is the improvement here? all I see is changing order of import

File: CHANGELOG.md [hunk 6]
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## [7.29.5] - 2024-07-16
 ### Fixed
 - fix invalid error about `[CryptoComparePriceMultiFetcher] empty fsyms`
+- fix PolygonIOCurrencySnapshotGramsFetcher and MetalPriceApiFetcher fetcher
```
Comment (DZariusz) [inline CHANGELOG.md:17]: this is wrong place

**Review by DZariusz — APPROVED** [review]
